### DISCUSSION TITLE
Should `ESC.Component` switch to `useEffect`?

### DISCUSSION THREAD

**Discussion by rechenberger** [hmans/miniplex#304] [Q&A]
Hi.

As described in issue #303, I had problems connecting `@react-three/rapier`'s `RigidBody` via:

```tsx
<ECS.Component name="rigidBody">
  <RigidBody />
<ECS.Component>
```

https://stackblitz.com/edit/vitejs-vite-dt2zjv?file=src/App.tsx
I got the example to work 🎉 by forking the `ESC.Component` from `createReactAPI`. The "fix" was to change the 2 `useIsomorphicLayoutEffect`s to `useEffect`.

This is fine for me as I a can continue exploring `miniplex`. Not really sure of what the implications of changing to `useEffect` are and if it would improve the library.

@hmans Any seconds thoughts?

**Comment by hmans** [comment]
Thanks for exploring this further. If changing to a `useEffect` fixes the issue, it probably means that the `ref` that `<RigidBody>` is forwarding is now being acquired at a different time in the pipeline than before.

I don't think using `useEffect`, even if it seems to solve the issue you're having, is a good solution, because you don't have a lot of control over when it runs, and with R3F, it will (if I remember correctly) sometimes (often? always?) even run _after_ the frame has rendered, which is probably not what you want and will lead to more issues further down the road.

But, granted, both my React and especially R3F are a little rusty since I haven't actively used them in quite some time.

Maybe I can humbly ping @wiledal to weigh in? 🙏

**Reply by wiledal** [reply]
One of the big hurdles is handling the creation and destruction of Rapier objects at "the right time". In `r3/rapier`, we try to set things up "as soon as possible", but this can lead to desynchronisation of references, as effects trigger from the inside out. The World must exist before a RigidBody, and the RigidBody must exist before the Collider. And when destroying a scene the World must exist until the last Collider has been removed 😅 

The recommended way to access RigidBody `refs` in `r3/rapier` right now is from within `useEffect`, as it should not create these problems. But the incompatibility that this brings to other libraries is new to me (and sad). I'm going to think about how this can be improved...
